lightning/src/ln/channelmanager.rs

```diff
@@ -442,6 +442,17 @@ pub struct ChannelManager<Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref,
 	/// Locked *after* channel_state.
 	pending_inbound_payments: Mutex<HashMap<PaymentHash, PendingInboundPayment>>,
 
+	/// The session_priv bytes of outbound payments which are pending resolution.
+	/// The authoritative of these HTLCs resides either within Channels or ChannelMonitors (if the
+	/// channel has been force-closed), however we track them here to prevent duplicative
+	/// PaymentSent/PaymentFailed events. Specifically, in the case of a duplicative
+	/// update_fulfill_htlc message after a reconnect, we may "claim" a payment twice.
+	/// Additionally, for channels which have been force-closed, because the ChannelMonitor event
+	/// is ultimately handled by us, and we aren't supposed to generate duplicative events unless we
+	/// haven't been re-serialized, we have to de-duplicate them here.
```

Comment:
I rewrote this section a bit, let me know if its better.

Reply:
Suggested phrasing: "Additionally, because we don't write out ChannelMonitors after each new block, it's possible that replaying blocks on ChannelMonitors after restart can result in duplicate `update_fulfill_htlc` events coming from force-closed ChannelMonitors."

Reply:
Ok, Rewrote it again more similar to what you suggested here. Note that it isn't actually update_fulfill_htlc messages/events.